### PR TITLE
Changes for 1.2.0 from LUMC GitLab

### LINKED ISSUE #41: Saving after connect_as_extension
As noted on the [Zemax forum](https://community.zemax.com/zos-api-12/tryind-to-get-started-with-zospy-4615?postid=14376#post14376), an error is currently raised when trying to use `OpticStudioSystem.Save` after connecting as extension. In this specific case, the private property `_OpenFile` is kept as None. 

I can see two ways to fix this:
1. Set the _OpenFile property directly after connecting as extension
2. Dropping the 'OpenFile' attribute all together and directly checking `OpticStudioSystem.SystemFile`. As this is a private property, I would not directly consider it a breaking change.

It might be good to address this when addressing https://github.com/MREYE-LUMC/ZOSPy/discussions/34

### PR DESCRIPTION
**Pull request by crnh**
## Proposed change

Merge internally developed changes for ZOSPy 1.2.0.

## Type of change

- [ ] Example (a notebook demonstrating how to use ZOSPy for a specific application)
- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New analysis (a wrapper around an OpticStudio analysis)
- [x] New feature (other than an analysis)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests
- [x] Documentation (improvement of either the docstrings or the documentation website)

# Additional information

- Python version: 3.9 - 3.12
- OpticStudio version: 2024 R1

## Related issues

Related discussions: #20, #34, #47, #57
Closes #41

## Checklist

- [x] I have followed the [contribution guidelines][contribution-guidelines]
- [x] The code has been linted, formatted and tested locally using tox.
- [x] Local tests pass. **Please fix any problems before opening a PR. If this is not possible, specify what doesn't work and why you can't fix it.**
- [x] I added new tests for any features contributed, or updated existing tests.
- [x] I updated CHANGELOG.md with my changes (except for refactorings and changes in the documentation).

If you contributed an analysis:

- [x] I did not use `AttrDict`s for the analysis result data (please use dataclasses instead).

If you contributed an example:

- [ ] I contributed my example as a Jupyter notebook.
    

[contribution-guidelines]: https://github.com/MREYE-LUMC/ZOSPy/blob/main/CONTRIBUTING.md
[unittest-instructions]: https://github.com/MREYE-LUMC/ZOSPy/blob/main/tests/README.md
[numpydoc]: https://numpydoc.readthedocs.io/en/latest/format.html

### CODE REVIEW

**Comment by crnh** [comment]
Tests are currently running!

File: tests/analyses/test_raysandspots.py [hunk 1]
```diff
@@ -112,12 +111,12 @@ def test_ray_fan_returns_correct_result(
         )
 
         assert np.allclose(
-            result.Data[f"Tangential fan, field number 1 = 0{DECIMAL_POINT}0000 (deg)"],
```
Comment (crnh) [inline tests/analyses/test_raysandspots.py:115]: This should not be reverted

Comment (crnh) [reply tests/analyses/test_raysandspots.py:115]: Other occurrences below

File: tests/conftest.py [hunk 2]
```diff
@@ -49,62 +56,31 @@ def system_save_file(request):
...
-    application = zos.Application
+def optic_studio_version(zos, connection_mode) -> Version:
+    zos.connect(connection_mode)
 
-    version = f"{application.ZOSMajorVersion}.{application.ZOSMinorVersion}.{application.ZOSSPVersion}"
+    version = Version(
+        major=zos.Application.ZOSMajorVersion, minor=zos.Application.ZOSMinorVersion, patch=zos.Application.ZOSSPVersion
+    )
 
     zos.Application.CloseApplication()
```
Comment (crnh) [inline tests/conftest.py:85]: Set `zos.Application` to `None` here

File: tests/test_zpcore.py [hunk 3]
```diff
@@ -137,6 +150,22 @@ def test_create_new_system_raises_valueerror(zos, simple_system, connection_mode
         zos.create_new_system()
 
 
+@pytest.fixture()
+@pytest.mark.require_mode("standalone")
```
Comment (crnh) [inline tests/test_zpcore.py:154]: This marker appears not to work and can be removed, see https://docs.pytest.org/en/latest/how-to/mark.html

File: zospy/api/__init__.py [hunk 4]
```diff
@@ -0,0 +1,8 @@
+"""ZOS-API connection management.
+
+This module contains functions for initialization of and communication with the ZOS-API.
+"""
+
+# from zospy.api import codecs
```
Comment (crnh) [inline zospy/api/__init__.py:6]: Remove commented code

File: zospy/zpcore.py [hunk 5]
```diff
@@ -222,14 +228,16 @@ def new(self, saveifneeded: bool = False):
 
         logger.info("Opened new file")
 
-    def save_as(self, filepath: str):
+    def save_as(self, filepath: str | Path):
```
Comment (crnh) [inline zospy/zpcore.py:231]: Use PathLike here

**Comment by crnh** [comment]
All tests pass!

**Review by jwmbeenakker — CHANGES_REQUESTED** [review]
The readthedocs is not up to date:
- FAQ: it still uses the legacy connection method
- Compatibility: support for python 3.12 is missing

File: README.md [hunk 6]
```diff
@@ -14,12 +14,13 @@ In addition to full access to all the OpticStudio fucntions through the ZOS-API,
 - Wrapper functions for several OpticStudio analyses in `zospy.analyses`;
 - Easy access to solvers in `zospy.solvers`;
 - Easy access to all API constants in `zospy.constants`;
-- Autocomplete for all ZOS-API endpoints and constants.
+- Autocomplete for all ZOS-API endpoints and constants;
+- Automatic downcasting of common ZOS-API interfaces to their implementation.
```
Comment (crnh) [inline README.md:18]: Do we want to mention this in the README? It requires a quite strong technical background to understand what this means, and I don't expect the average user to link this to Python.NET 3 related problems

**Review by jwmbeenakker — APPROVED** [review]
all good to go

**Pull request merged**